lib/martian_robot.rb

```diff
@@ -13,24 +13,50 @@ def initialize(coordinates, orientation, planet)
   end
 
   def report_final_position(instruction)
-    final_position = starting_position
-    instruction.split('').each {|single_instruction| execute(single_instruction, final_position)}
-    lost?(final_position) ? :lost : final_position
+    instruction.split('').each {|single_instruction|
+      execute(single_instruction)
+      break if lost_before_executing_all_instruction(@coordinates)
+    }
+    if lost?(@coordinates)
+      stop_or_get_lost
+    else
+      @coordinates
+    end
   end
 
-
-  def starting_position
-    x = @coordinates[0]
-    y = @coordinates[1]
-    [x,y]
+  def last_position_before_lost
+    last_position = [@coordinates[0], @coordinates[1]]
+    if @coordinates[0] > @planet.size
+      last_position[0] = @planet.size
+    elsif @coordinates[1] > @planet.size
+      last_position[1] = @planet.size
+    elsif @coordinates[0] < 0
+      last_position[0] = 0
+    else
+      last_position[1] = 0
+    end
+    last_position
   end
 
   private
 
-  def execute(instruction, position)
+  def lost_before_executing_all_instruction(current_position)
```

Comment:
Without it, a robot that gets outside the confines and returns back eventually would report its position instead of returning :lost. That method checks the robot position after each iteration and if it gets lost before all the iterations are made, the loop breaks. We can discuss this later!